src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java

```diff
@@ -1405,18 +1397,61 @@ public void cleanupExam(Long examId, Principal principal) {
         final var auditEvent = new AuditEvent(principal.getName(), Constants.CLEANUP_EXAM, "exam=" + examId);
         auditEventRepository.add(auditEvent);
 
-        // The Objects::nonNull is needed here because the relationship exam -> exercise groups is ordered and
-        // hibernate sometimes adds nulls into the list of exercise groups to keep the order
-        Set<Exercise> examExercises = examRepository.findByIdWithExamUsersExerciseGroupsAndExercisesElseThrow(examId).getExerciseGroups().stream().filter(Objects::nonNull)
-                .map(ExerciseGroup::getExercises).flatMap(Collection::stream).collect(Collectors.toSet());
+        var programmingExercises = getAllExercisesForExamByType(examId, ProgrammingExercise.class);
+        programmingExercises.forEach(exercise -> exerciseDeletionService.cleanup(exercise.getId(), true));
+        log.info("The exam {} has been cleaned up!", examId);
+    }
 
-        examExercises.forEach(exercise -> {
-            if (exercise instanceof ProgrammingExercise) {
-                exerciseDeletionService.cleanup(exercise.getId(), true);
+    /**
+     * Updates the working times for student exams based on a given change in working time and reschedules exercises accordingly.
+     * This method considers any existing time extensions for individual students and adjusts their working times relative to the original exam duration and the specified change.
+     * After updating the working times, it saves the changes and, if the exam is already visible, notifies both the students and relevant instances about the update.
+     * Additionally, if the current time is before the latest individual exam end date, it potentially triggers a rescheduling of the clustering of modeling submissions,
+     * considering the use of Compass.
+     *
+     * @param exam                 The exam entity for which the student exams and exercises need to be updated and rescheduled.
+     * @param originalExamDuration The original duration of the exam, in minutes, before any changes.
+     * @param workingTimeChange    The amount of time, in minutes, to add or subtract from the exam's original duration and the student's working time. This value can be positive
+     *                                 (to extend time) or negative (to reduce time).
+     */
+    public void updateStudentExamsAndRescheduleExercises(Exam exam, Integer originalExamDuration, Integer workingTimeChange) {
+        var now = now();
+
+        User instructor = userRepository.getUser();
+
+        var studentExams = studentExamRepository.findByExamId(exam.getId());
```

Comment:
Do we need to do a db call here? Can we maybe fetch them in the original Exam call(s)?